File: 3Sum Closest.cpp
// Leetcode Problem 16: 3Sum Closest
// C++ CODE
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int n = nums.size();
        int closestSum = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < n - 2; i++) {
            int left = i + 1;
            int right = n - 1;

            while (left < right) {
                int currSum = nums[i] + nums[left] + nums[right];

                if (abs(target - currSum) < abs(target - closestSum)) {
                    closestSum = currSum;
                }

                if (currSum < target) {
                    left++;
                } 
                else if (currSum > target) {
                    right--;
                } 
                else {
                    return currSum; // Exact match found
                }
            }
        }

        return closestSum;
    }
};

File: 3Sum Closest.py
# Leetcode Problem 16: 3Sum Closest
# Python CODE
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        closest = nums[0] + nums[1] + nums[2]

        for i in range(n - 2):
            left = i + 1
            right = n - 1

            while left < right:
                sum = nums[i] + nums[left] + nums[right]

                if abs(sum - target) < abs(closest - target):
                    closest = sum

                if sum == target:
                    return sum
                elif sum < target:
                    left += 1
                else:
                    right -= 1

        return closest